File: screen-stream/screen-stream/FirstViewController.m
//
//  FirstViewController.m
//  screen-stream
//
//  Created by xiang on 13/09/2017.
//  Copyright © 2017 dotEngine. All rights reserved.
//

#import "FirstViewController.h"


#import <SceneKit/SceneKit.h>
#import <ARKit/ARKit.h>

#import <DotEngine.h>
#import <DotStream.h>
#import <DotVideoCapturer.h>


#import "CVPixelBufferResize.h"

static  NSString*  APP_KEY = @"45";
static  NSString*  APP_SECRET = @"dc5cabddba054ffe894ba79c2910866c";
static  NSString*  ROOM = @"screen_test";


@interface FirstViewController ()<DotEngineDelegate,DotStreamDelegate,ARSCNViewDelegate, ARSessionDelegate>
{
    
    DotEngine* dotEngine;
    DotStream* localStream;
    DotVideoCapturer* videoCapturer;
    BOOL isStarted;
    
    NSDate* lastDate;
    CVPixelBufferResize* resize;
    
    CADisplayLink *displayLink;
}


@property (weak, nonatomic) IBOutlet ARSCNView *sceneView;

@property (nonatomic, strong) SCNRenderer *renderer;
@property (nonatomic, strong) dispatch_queue_t videoQueue;
@property (nonatomic, assign) CGSize outputSize;

@end

@implementation FirstViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    self.sceneView.delegate = self;
    
    self.sceneView.showsStatistics = YES;
    
    SCNScene *scene = [SCNScene sceneNamed:@"art.scnassets/ship.scn"];
    
    self.sceneView.scene = scene;
    self.sceneView.session.delegate = self;
    
    self.renderer = [SCNRenderer rendererWithDevice:nil options:nil];
    self.renderer.scene = scene;
    
    self.videoQueue = dispatch_queue_create("cc.dot.video.queue", NULL);
    
    
    dotEngine = [DotEngine sharedInstanceWithDelegate:self];
    localStream = [[DotStream alloc] initWithAudio:YES
                                             video:YES
                                      videoProfile:DotEngine_VideoProfile_480P
                                          delegate:self];
    
    videoCapturer = [[DotVideoCapturer alloc] init];
    localStream.videoCaptuer = videoCapturer;
    
    resize = [[CVPixelBufferResize alloc] init];
    
    [UIApplication sharedApplication].idleTimerDisabled = YES;
    
    self.outputSize = self.sceneView.frame.size;
    
}


-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    ARWorldTrackingConfiguration *configuration = [ARWorldTrackingConfiguration new];
    // Run the view's session
    [self.sceneView.session runWithConfiguration:configuration];
}


- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    // Pause the view's session
    [self.sceneView.session pause];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (IBAction)startCapture:(id)sender {
    // startCaptureWithHandler
    
    if (isStarted) {
        return;
    }
    
    uint32_t randomNum = arc4random_uniform(10000);
    
    NSString* userId = [NSString stringWithFormat:@"stream%d",randomNum];
    
    [localStream setupLocalMedia];
    
    [dotEngine generateTestTokenWithAppKey:APP_KEY
                                 appsecret:APP_SECRET
                                      room:ROOM
                                    userId:userId
                                 withBlock:^(NSString *token, NSError *error) {
                                     
                                     if (token != nil) {
                                         [dotEngine joinRoomWithToken:token];
                                     }
                                    }];
    
    
    displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(getCaptureData)];
    displayLink.preferredFramesPerSecond =  15;  // 每秒15帧 可以自己调整
    [displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
    
}

- (IBAction)stopCapture:(id)sender {
    
    //stopCaptureWithHandler
    if (!isStarted) {
        return;
    }
    if (displayLink!= nil) {
        [displayLink invalidate];
        displayLink = nil;
    }
    isStarted = false;
}


-(CVPixelBufferRef)capturePixelBuffer {
    
    UIImage *image = [self.renderer snapshotAtTime:1 withSize:CGSizeMake(self.outputSize.width, self.outputSize.height) antialiasingMode:SCNAntialiasingModeMultisampling4X];
    
    CIImage* ciimage = [[CIImage alloc] initWithCGImage:image.CGImage];
    
    CVPixelBufferRef  pixelBuffer = [resize processCIImage:ciimage];
    return pixelBuffer;
}


-(void)getCaptureData
{
    NSLog(@"getCaptureData ");
    
    dispatch_async(self.videoQueue, ^{
        CVPixelBufferRef pixelBuffer = [self capturePixelBuffer];
        if (pixelBuffer) {
            [videoCapturer sendCVPixelBuffer:pixelBuffer rotation:VideoRoation_0];
             CFRelease(pixelBuffer);
        } else {
            NSLog(@"can not get pixel buffer ");
        }
        
    });
}

#pragma DotEngine Delegate


-(void)dotEngine:(DotEngine* _Nonnull) engine didJoined:(NSString* _Nonnull)peerId
{
    NSLog(@"didJoined ");
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didLeave:(NSString* _Nonnull)peerId
{
    NSLog(@"didLeave");
}

-(void)dotEngine:(DotEngine* _Nonnull) engine  stateChange:(DotStatus)state
{
    if (state == DotStatusConnected) {
        //[dotEngine addStream:localStream];
    }
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didAddLocalStream:(DotStream* _Nonnull)stream
{
    NSLog(@"didAddLocalStream");
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didRemoveLocalStream:(DotStream* _Nonnull)stream
{
    
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didAddRemoteStream:(DotStream* _Nonnull)stream
{
    
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didRemoveRemoteStream:(DotStream* _Nonnull) stream
{
    
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didOccurError:(DotEngineErrorCode)errorCode
{
    
    NSLog(@"didOccurError  %ld", (long)errorCode);
}


#pragma DotStream delegate


-(void)stream:(DotStream* _Nullable)stream  didMutedVideo:(BOOL)muted
{
    
}

-(void)stream:(DotStream* _Nullable)stream  didMutedAudio:(BOOL)muted
{
    
}

-(void)stream:(DotStream* _Nullable)stream  didGotAudioLevel:(int)audioLevel
{
    
}


#pragma  ARSessionDelegate


- (void)session:(ARSession *)session didUpdateFrame:(ARFrame *)frame
{
    NSLog(@"didUpdateFrame ");
}

- (void)session:(ARSession *)session didAddAnchors:(NSArray<ARAnchor*>*)anchors
{
    
    
}


- (void)session:(ARSession *)session didUpdateAnchors:(NSArray<ARAnchor*>*)anchors
{
    
    
}


- (void)session:(ARSession *)session didRemoveAnchors:(NSArray<ARAnchor*>*)anchors
{
    
}

@end

File: screen-stream/screen-stream-swift/FirstViewController.swift
//
//  FirstViewController.swift
//  screen-stream-swift
//
//  Created by Gong Zhang on 2017/9/15.
//  Copyright © 2017年 dotEngine. All rights reserved.
//

import UIKit
import ReplayKit

let APP_KEY = "45"
let APP_SECRET = "dc5cabddba054ffe894ba79c2910866c"
let ROOM = "screen_stream"

class FirstViewController: UIViewController, RPScreenRecorderDelegate, DotEngineDelegate, DotStreamDelegate {
    
    @IBOutlet weak var timeLabel: UILabel!
    
    var screenRecorder: RPScreenRecorder!
    var dotEngine: DotEngine!
    var localStream: DotStream!
    var videoCapturer: DotVideoCapturer?
    
    var isStarted: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        screenRecorder = RPScreenRecorder.shared()
        screenRecorder.delegate = self
        
        dotEngine = DotEngine.sharedInstance(with: self)
        localStream = DotStream(audio: true, video: true, videoProfile: DotEngineVideoProfile.DotEngine_VideoProfile_480P, delegate: self)
        
        videoCapturer = DotVideoCapturer()
        localStream.videoCaptuer = videoCapturer
        
        let formatter = DateFormatter()
        formatter.dateFormat = "hh:mm A"
        Timer.scheduledTimer(withTimeInterval: 0.2, repeats: true) { [weak self] t in
            guard let me = self else {
                t.invalidate()
                return
            }
            me.timeLabel.text = formatter.string(from: Date())
        }
    }
    
    @IBAction func buttonStartAction(_ sender: Any) {
        guard !isStarted else {
            return
        }
        
        let randomNum = arc4random_uniform(10000)
        let userId = "stream\(randomNum)"
        
        localStream.setupLocalMedia()
        
        dotEngine.generateTestToken(withAppKey: APP_KEY, appsecret: APP_SECRET, room: ROOM, userId: userId) { [weak self] (token, err) in
            guard err == nil else {
                print(err!)
                return
            }
            
            guard let token = token else {
                print("token is nil")
                return
            }
            
            self?.dotEngine.joinRoom(withToken: token)
        }
        
        screenRecorder.startCapture(handler: { (sampleBuffer, bufferType, err) in
            
            // TODO: 应该缩放一下屏幕
            // TODO: 目前先只支持视频
            
            if bufferType == .video {
                let imageBuffer = CMSampleBufferGetImageBuffer(sampleBuffer)!
                CVPixelBufferLockBaseAddress(imageBuffer, [])
                
                let pixelFormatType = CVPixelBufferGetPixelFormatType(imageBuffer)
                
                if pixelFormatType == kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange ||
                    pixelFormatType == kCVPixelFormatType_420YpCbCr8BiPlanarFullRange {
                    // only nv12 support
                    print("kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange")
                    
                    if let videoCapture = self.videoCapturer {
                        videoCapture.send(imageBuffer, rotation: .roation_0)
                    }
                    
                }
                
                CVPixelBufferUnlockBaseAddress(imageBuffer, [])
            }
            
        }) { [weak self] err in
            if err == nil {
                self?.isStarted = true
            } else {
                print(err!)
            }
        }
    }
    
    @IBAction func buttonStopAction(_ sender: Any) {
        guard isStarted else {
            return
        }
        
        screenRecorder.stopCapture { [weak self] err in
            if let err = err {
                print(err)
            }
            self?.isStarted = false
        }
    }
    
    // MARK: - RPScreenRecorder Delegates
    
    func screenRecorderDidChangeAvailability(_ screenRecorder: RPScreenRecorder) {
        print("screenRecorderDidChangeAvailability \(screenRecorder.isAvailable)")
    }
    
    // MARK: - DotEngine Delegates
    
    func dotEngine(_ engine: DotEngine, didJoined peerId: String) {
        print("didJoined")
    }
    
    func dotEngine(_ engine: DotEngine, didLeave peerId: String) {
        print("didLeave")
    }
    
    func dotEngine(_ engine: DotEngine, stateChange state: DotStatus) {
        if state == .connected {
            dotEngine.add(localStream)
        }
    }
    
    func dotEngine(_ engine: DotEngine, didAddLocalStream stream: DotStream) {
        print("didAddLocalStream")
    }
    
    func dotEngine(_ engine: DotEngine, didRemoveLocalStream stream: DotStream) {
    }
    
    func dotEngine(_ engine: DotEngine, didAddRemoteStream stream: DotStream) {
    }
    
    func dotEngine(_ engine: DotEngine, didRemoveRemoteStream stream: DotStream) {
    }
    
    func dotEngine(_ engine: DotEngine, didOccurError errorCode: DotEngineErrorCode) {
        print("didOccurError \(errorCode)")
    }
    
    // MARK: - DotStream Delegates
    
    func stream(_ stream: DotStream?, didMutedVideo muted: Bool) {
    }
    
    func stream(_ stream: DotStream?, didMutedAudio muted: Bool) {
    }
    
    func stream(_ stream: DotStream?, didGotAudioLevel audioLevel: Int32) {
    }
    
}
